modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -824,7 +824,28 @@ def from_labels(self) -> "PandasDataframe":
         def from_labels_executor(df, **kwargs):
             # Setting the names here ensures that external and internal metadata always match.
             df.index.names = new_column_names
-            return df.reset_index()
+
+            # Implementation of handling case when columns have the same
+            # name as one of index levels names via `insert` (other option is use
+            # lazy_metadata_decorator(apply_axis="both"))
+            if any(level_name in df.columns for level_name in df.index.names):
+                if df.index.nlevels > 1:
+                    columns_to_insert = df.index.to_frame().reset_index(drop=True)
+                    df = df.reset_index(drop=True)
+                    for col in columns_to_insert.columns[::-1]:
+                        df.insert(0, col, columns_to_insert[col], allow_duplicates=True)
+                else:
+                    columns_to_insert = df.index
+                    df = df.reset_index(drop=True)
+                    df.insert(
+                        0,
+                        columns_to_insert.name,
+                        columns_to_insert,
+                        allow_duplicates=True,
+                    )
+                return df
+            else:
+                return df.reset_index()
```

Comment:
I think we don't need the extra checks, we should just be able to handle all of the cases in the same way here.
```suggestion
            columns_to_insert = df.index.to_frame().reset_index(drop=True)
            df = df.reset_index(drop=True)
            pandas.concat([columns_to_insert, df], axis=1)
```

Reply:
I think that your suggestion should be under `if any(level_name in df.columns for level_name in df.index.names):` condition, because it will be not efficient to use those number of operations for each `from_labels_executor` call. This idea is implemented in 10c2c99240eb94f90b12bfeea480934b35ee75de.